### PR TITLE
Implementing SymmetricDifference class, #8179

### PR DESCRIPTION
**Pull request by kaichogami**
Created a class to calculate the Symmetric Difference of two finite sets(see #8179) . This class structure is similar to that of already implemented `Complement` class. `__xor__` method in `FiniteSet` class is overridden to call `SymmetricDifference` class. 
Usage Example

```
In [1]: from sympy import FiniteSet, SymmetricDifference
In [2]: SymmetricDifference(FiniteSet(1,2,3,4), FiniteSet(3,4,5))
Out[2]: {1, 2, 5}
```

### CODE REVIEW

**Comment by kaichogami** [comment]
@pelegm I added `NotImplementedError` to `Set`, `Interval`,  `UniversalSet`, `EmptySet` classes. `UniversalSet`, `EmptySet` should just return `other`, although I think as it is not in the scope of the issue #8179, we should create a new issue or directly a PR to make the changes in the above four classes.
Also if you think I missed something, let me know!

**Comment by hargup** [comment]
>  I added `NotImplementedError` to `Set`, `Interval`,  `UniversalSet`, `EmptySet` classes. `UniversalSet`, `EmptySet` should just return `other`, although I think as it is not in the scope of the issue #8179

Extending the PR to handle other Set classes shouldn't be difficult if you use the `Complement` and `Union` classes to represent Symmetric Difference. See https://en.wikipedia.org/wiki/Symmetric_difference#Properties (this was also mention by @pelegm in https://github.com/sympy/sympy/issues/8179#issuecomment-58139786). You don't need to write the logic to actually evaluate the Symmmetric Difference. The reduce functions in `Complement` and `Union` classes can handle that.

File: sympy/sets/tests/test_sets.py [hunk 1]
```diff
@@ -831,3 +831,12 @@ def test_Eq():
 
     assert Eq(s1*s2, s1*s2)
     assert Eq(s1*s2, s2*s1) == False
+
+def test_SymmetricDifference():
+   assert SymmetricDifference(FiniteSet(0,1,2,3,4,5), FiniteSet(2,4,6,8,10)) == FiniteSet(0, 1, 3, 5, 6, 8, 10)
+   assert SymmetricDifference(FiniteSet(2,3,4), FiniteSet(2,3,4,5)) == FiniteSet(5)
+   assert FiniteSet(1,2,3,4,5) ^ FiniteSet(1,2,5,6) == FiniteSet(3,4,6)
+
+   #Not Implemented yet
```
Comment (hargup) [inline sympy/sets/tests/test_sets.py]: You can create a test function with the `@xfail` decorator. `xfail` implies `expected to fail`.

**Comment by kaichogami** [comment]
@hargup using the classes made it easy. If you think I still missed something, let me know!

**Comment by asmeurer** [comment]
Returning an unevaluated object is better than raising NotImplementedError.

File: sympy/sets/sets.py [hunk 2]
```diff
@@ -206,6 +206,12 @@ def _complement(self, other):
         elif isinstance(other, FiniteSet):
             return FiniteSet(*[el for el in other if self.contains(el) != True])
 
+    def _symmetricdifference(self, other):
+        """
+        This function should only be used internally
```
Comment (skirpichev) [inline sympy/sets/sets.py]: I think, you can remove this docstring.  Method is private if it starts with underscore.

File: sympy/sets/sets.py [hunk 3]
```diff
@@ -882,6 +891,13 @@ def _complement(self, other):
         return Set._complement(self, other)
 
 
+    def _symmetricdifference(self, other):
```
Comment (skirpichev) [inline sympy/sets/sets.py]: This is exactly same method as for Set.  Remove?

File: sympy/sets/sets.py [hunk 4]
```diff
@@ -1716,6 +1738,12 @@ def _union(self, other):
 
         return None
 
+    def _symmetricdifference(self, other):
+        """
+        This function should only be used internally
+        """
+        return Union(Complement(self, other), Complement(other, self))
```
Comment (skirpichev) [inline sympy/sets/sets.py]: again, same as for Set

File: sympy/sets/tests/test_sets.py [hunk 5]
```diff
@@ -831,3 +831,15 @@ def test_Eq():
 
     assert Eq(s1*s2, s1*s2)
     assert Eq(s1*s2, s2*s1) == False
+
```
Comment (skirpichev) [inline sympy/sets/tests/test_sets.py]: add newline

File: sympy/sets/sets.py [hunk 6]
```diff
@@ -206,6 +206,12 @@ def _complement(self, other):
         elif isinstance(other, FiniteSet):
             return FiniteSet(*[el for el in other if self.contains(el) != True])
 
+    def _symmetricdifference(self, other):
```
Comment (skirpichev) [inline sympy/sets/sets.py]: Probably, we should have a pair of private and public methods, just like other cases.  Also, please see documentation of [builtin set](https://docs.python.org/3/library/stdtypes.html#set-types-set-frozenset) class.  I think, that's why public method should be named as `symmetric_difference`.

File: sympy/sets/sets.py [hunk 7]
```diff
@@ -1808,6 +1836,43 @@ def __lt__(self, other):
         return self.is_proper_subset(other)
 
 
+class SymmetricDifference(Set):
+    """Representes the Symmetric Difference of a set with another set.
```
Comment (skirpichev) [inline sympy/sets/sets.py]: Represents

Comment (skirpichev) [inline sympy/sets/sets.py]: You could add reference here, i.e. on the wiki page.

File: sympy/sets/sets.py [hunk 8]
```diff
@@ -1808,6 +1836,43 @@ def __lt__(self, other):
...
+
+    Examples
+    ========
+
+    >>> from sympy import SymmetricDifference, FiniteSet
+    >>> SymmetricDifference(FiniteSet(1, 2, 3), FiniteSet(3, 4, 5))
+    {1, 2, 4, 5}
+
+    See Also
+    ========
```
Comment (skirpichev) [inline sympy/sets/sets.py]: add newline after section header, same below

File: sympy/sets/sets.py [hunk 9]
```diff
@@ -1808,6 +1836,43 @@ def __lt__(self, other):
...
+
+    def __new__(cls, a, b, evaluate=True):
+        if evaluate:
+            return SymmetricDifference.reduce(a, b)
+
+        return Basic.__new__(cls, a, b)
+
+    @staticmethod
+    def reduce(A, B):
+
```
Comment (skirpichev) [inline sympy/sets/sets.py]: remove this newline

File: sympy/sets/sets.py (hunk 7 above)
Comment (kaichogami) [inline sympy/sets/sets.py]: @skirpichev Could you tell me what you mean by "You could add reference here, i.e. on the wiki page."? I added the definition from the wikipedia page in docstring.

**Comment by kaichogami** [comment]
@skirpichev I made the said changes. I have still lot to learn. 
This looks all set to go to me now.

File: sympy/sets/sets.py (hunk 7 above)
Comment (skirpichev) [inline sympy/sets/sets.py]: Yes, but you could add citation of this page right here (with `[1]_` in an appropriate place).

Comment (skirpichev) [inline sympy/sets/sets.py]: See also [this](https://github.com/sympy/sympy/wiki/Writing-documentation) about writing documentation.

**Comment by kaichogami** [comment]
@skirpichev @pelegm @hargup Looks like its done now.

**Comment by skirpichev** [comment]
+1

But please fix merge conflicts.  Also, please squash all commits (except the first) and edit commit message: you can do this with "git rebase".

**Comment by kaichogami** [comment]
@skirpichev I squashed the commits, except the first one, into a single commit after fixing the merge conflicts. For what reason are squashing the commits?

**Comment by skirpichev** [comment]
I still see two commits.

> For what reason are we squashing the commits?

To cleanup the git history.  Also,  "first commit to solve bug #8179" is not a [good commit message](https://github.com/sympy/sympy/wiki/Development-workflow#writing-commit-messages).

File: sympy/sets/sets.py [hunk 10]
```diff
@@ -206,6 +206,12 @@ def _complement(self, other):
         elif isinstance(other, FiniteSet):
             return FiniteSet(*[el for el in other if self.contains(el) != True])
 
+    def symmetric_difference(self, other):
+         return SymmetricDifference(self, other)
+
+    def _symmetricdifference(self, other):
```
Comment (skirpichev) [inline sympy/sets/sets.py]: probably, this name should match public method?

**Comment by kaichogami** [comment]
@skirpichev I squashed all the commits into one and wrote a better commit message. Looks much cleaner now.

**Comment by hargup** [comment]
+1, will merge in 24h if no objection. @kaichogami you should also add printer for the `SymmetricDifference` class but you can save that for another PR.

**Comment by kaichogami** [comment]
@hargup alright. I'll create a new PR as soon as I add the printer.

**Comment by hargup** [comment]
Thanks it is in.

**Pull request merged**